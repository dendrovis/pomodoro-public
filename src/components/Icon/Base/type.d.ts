type Theme = 'light' | 'dark';
export type State = {
  [id: string]: {
    elementContainer?: HTMLDivElement;
    isClosed: boolean;
    icon: string;
    theme: Theme;
  };
};
export type CreateContainer = (id: string, onClick: IsClosed) => HTMLDivElement;
export type SetTheme = (id: string, theme: Theme) => void;
export type SetClosed = (id: string, isClosed: boolean) => void;
export type IsClosed = () => void;

export interface BaseType {
  (params: {
    id: string;
    icon: string;
    onClick: IsClosed;
    theme?: Theme;
    elementTargetID: string;
  }): void;
  setTheme: SetTheme;
  setClosed: SetClosed;
}
