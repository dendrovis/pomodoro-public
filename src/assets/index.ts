export { default as ICON_ALARM_OFF } from './images/static/alarm_off.svg?raw';
export { default as ICON_ALARM_ON } from './images/static/alarm_on.svg?raw';
export { default as ICON_CACHE_OFF } from './images/static/cache_off.svg?raw';
export { default as ICON_CACHE_ON } from './images/static/cache_on.svg?raw';
export { default as IMG_LIGHT_DARK } from './images/static/light_dark_background.webp';
export { default as IMG_SUN } from './images/static/sun.webp';
export { default as IMG_MOON } from './images/static/moon.webp';
export { default as ICON_LOCK } from './images/static/lock.svg?raw';
export { default as ICON_UNLOCK } from './images/static/unlock.svg?raw';
export { default as ICON_MINUS } from './images/static/minus.svg?raw';
export { default as ICON_PLUS } from './images/static/plus.svg?raw';
export { default as ICON_PLAY } from './images/static/play.svg?raw';
export { default as ICON_PAUSE } from './images/static/pause.svg?raw';
export { default as ICON_STOP } from './images/static/stop.svg?raw';
export { default as ANIMATED_CHARACTER_IDLE } from './images/animated/idle.webp';
export { default as ANIMATED_CHARACTER_RUN } from './images/animated/run.webp';
export { default as ICON_CLOSE } from './images/static/close.svg?raw';
export { default as ICON_INFO } from './images/static/info.svg?raw';
export { default as ICON_HELP } from './images/static/help.svg?raw';
export { default as ICON_LOGO } from './images/static/logo.svg';
export { default as IMG_DENDROVIS } from './images/static/logo_dendrovis.webp';
