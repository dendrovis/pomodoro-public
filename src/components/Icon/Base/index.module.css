.icon-base {
  cursor: pointer;
  transition: all 0.3s ease-in-out;

  &:hover {
    transform: scale(1.5);
  }

  & > svg > path {
    cursor: pointer;
    fill: var(--tertiary-color-dark, black);
    transition: all 0.3s ease-in-out;
  }
}

.icon-base-closed {
  & > svg > path {
    cursor: pointer;
    fill: var(--secondary-color, red);
  }
}

.icon-base-dark {
  & > svg > path {
    cursor: pointer;
    fill: var(--primary-color, white);
  }
}
