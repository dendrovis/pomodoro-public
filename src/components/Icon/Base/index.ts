import { ICON_CLOSE } from '../../../assets';
import { isSVGRawTag } from '../../../utils';
import Classes from './index.module.css';
import { BaseType, CreateContainer, SetClosed, SetTheme, State } from './type';

const states: State = {};

const createContainer: CreateContainer = (id, onClick) => {
  const element = document.createElement('div');
  element.id = `icon-base-${id}`;
  element.onclick = () => onClick();
  element.className = `${Classes['icon-base']}`;
  return element;
};

const setTheme: SetTheme = (id, theme) => {
  const element = states[id].elementContainer;
  if (!element) throw new Error('element not available');
  if (theme === 'dark') {
    element.className = `${Classes['icon-base']} ${Classes['icon-base-dark']}`;
    states[id].theme = 'dark';
  } else {
    element.className = `${Classes['icon-base']}`;
    states[id].theme = 'light';
  }
};

const Base: BaseType = ({ id, icon, onClick, theme = 'light', elementTargetID }) => {
  if (Object.hasOwn(states, id))
    throw new Error('existed property, make sure property name is distinct');
  if (!isSVGRawTag(icon)) throw new Error('only support svg format');
  const elementTarget = document.querySelector(elementTargetID);
  if (!elementTarget) throw new Error('element not available');
  const elementContainer = createContainer(id, onClick);
  states[id] = { elementContainer, isClosed: false, icon, theme };
  elementContainer.innerHTML = icon;
  setTheme(id, theme);
  elementTarget.appendChild(elementContainer);
};

const setClosed: SetClosed = (id, isClosed) => {
  states[id].isClosed = isClosed;
  const element = states[id].elementContainer;
  if (!element) throw new Error('element not available');
  if (isClosed) {
    element.innerHTML = ICON_CLOSE;
    element.className = `${Classes['icon-base']} ${Classes['icon-base-closed']}`;
  } else {
    element.innerHTML = states[id].icon;
    setTheme(id, states[id].theme);
  }
};

Base.setTheme = setTheme;
Base.setClosed = setClosed;

export default Base;
