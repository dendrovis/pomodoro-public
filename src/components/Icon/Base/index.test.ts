import { describe, it, beforeEach, expect } from 'vitest';
import Base from '.';
import { ICON_LOCK, IMG_SUN } from '../../../assets';

beforeEach(() => {
  document.body.innerHTML = '';
  document.body.id = 'app';
});

describe('when given id', () => {
  it('should throw error if same id is used', () => {
    Base({
      id: 'test',
      icon: `<svg id="test"></svg>\n`,
      onClick: () => {},
      theme: 'light',
      elementTargetID: '#app',
    });
    expect(() =>
      Base({
        id: 'test',
        icon: `<svg id="test"></svg>\n`,
        onClick: () => {},
        theme: 'light',
        elementTargetID: '#app',
      })
    ).toThrowError();
  });
});

describe('when user click', () => {
  it('should invoke onClick callback', () => {
    let called = false;
    Base({
      id: 'test-2',
      icon: `<svg id="test"></svg>\n`,
      onClick: () => {
        called = true;
      },
      theme: 'light',
      elementTargetID: '#app',
    });
    const element = document.querySelector('#icon-base-test-2') as HTMLDivElement;
    element.click();
    expect(called).toBe(true);
  });
});

describe('when given icon', () => {
  it('should throw if not svg raw format', () => {
    expect(() =>
      Base({
        id: 'test-3',
        icon: IMG_SUN,
        onClick: () => {},
        theme: 'light',
        elementTargetID: '#app',
      })
    ).toThrowError();
  });
  it('should render svg raw format', () => {
    Base({
      id: 'test-4',
      icon: ICON_LOCK,
      onClick: () => {},
      theme: 'light',
      elementTargetID: '#app',
    });
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('when set theme', () => {
  it('should render the right theme', () => {
    Base({
      id: 'test-5',
      icon: ICON_LOCK,
      onClick: () => {},
      theme: 'light',
      elementTargetID: '#app',
    });
    expect(document.body.innerHTML).toMatchSnapshot();
    Base({
      id: 'test-6',
      icon: ICON_LOCK,
      onClick: () => {},
      theme: 'dark',
      elementTargetID: '#app',
    });
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('when set elementTargetID', () => {
  it('should throw error if element id does not exist', () => {
    expect(() =>
      Base({
        id: 'test-7',
        icon: ICON_LOCK,
        onClick: () => {},
        theme: 'light',
        elementTargetID: 'app',
      })
    ).toThrowError();
  });
  it('should render if element id exist', () => {
    Base({
      id: 'test-8',
      icon: ICON_LOCK,
      onClick: () => {},
      theme: 'light',
      elementTargetID: '#app',
    });
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('when icon is set close-able', () => {
  it('should render closed icon when clicked', () => {
    Base({
      id: 'test-9',
      icon: `<svg id="test"></svg>\n`,
      onClick: () => {},
      theme: 'light',
      elementTargetID: '#app',
    });
    Base.setClosed('test-9', true);
    const element = document.querySelector('#icon-base-test-9') as HTMLDivElement;
    element.click();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  it('should render origin icon when double clicked', () => {
    Base({
      id: 'test-10',
      icon: `<svg id="test"></svg>\n`,
      onClick: () => {},
      theme: 'dark',
      elementTargetID: '#app',
    });
    Base.setClosed('test-9', false);
    const element = document.querySelector('#icon-base-test-10') as HTMLDivElement;
    element.click();
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
