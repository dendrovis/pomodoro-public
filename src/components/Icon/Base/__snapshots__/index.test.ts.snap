// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when given icon > should render svg raw format 1`] = `
"<div id="icon-base-test-4" class="_icon-base_2178eb"><svg width="16" height="21" viewBox="0 0 16 21" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M2 21C1.45 21 0.979167 20.8042 0.5875 20.4125C0.195833 20.0208 0 19.55 0 19V9C0 8.45 0.195833 7.97917 0.5875 7.5875C0.979167 7.19583 1.45 7 2 7H3V5C3 3.61667 3.4875 2.4375 4.4625 1.4625C5.4375 0.4875 6.61667 0 8 0C9.38333 0 10.5625 0.4875 11.5375 1.4625C12.5125 2.4375 13 3.61667 13 5V7H14C14.55 7 15.0208 7.19583 15.4125 7.5875C15.8042 7.97917 16 8.45 16 9V19C16 19.55 15.8042 20.0208 15.4125 20.4125C15.0208 20.8042 14.55 21 14 21H2ZM2 19H14V9H2V19ZM8 16C8.55 16 9.02083 15.8042 9.4125 15.4125C9.80417 15.0208 10 14.55 10 14C10 13.45 9.80417 12.9792 9.4125 12.5875C9.02083 12.1958 8.55 12 8 12C7.45 12 6.97917 12.1958 6.5875 12.5875C6.19583 12.9792 6 13.45 6 14C6 14.55 6.19583 15.0208 6.5875 15.4125C6.97917 15.8042 7.45 16 8 16ZM5 7H11V5C11 4.16667 10.7083 3.45833 10.125 2.875C9.54167 2.29167 8.83333 2 8 2C7.16667 2 6.45833 2.29167 5.875 2.875C5.29167 3.45833 5 4.16667 5 5V7Z" fill="#FEF7F7"></path>
</svg>
</div>"
`;

exports[`when icon is set close-able > should render closed icon when clicked 1`] = `
"<div id="icon-base-test-9" class="_icon-base_2178eb _icon-base-closed_2178eb"><svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M6.4 15L10 11.4L13.6 15L15 13.6L11.4 10L15 6.4L13.6 5L10 8.6L6.4 5L5 6.4L8.6 10L5 13.6L6.4 15ZM10 20C8.61667 20 7.31667 19.7375 6.1 19.2125C4.88333 18.6875 3.825 17.975 2.925 17.075C2.025 16.175 1.3125 15.1167 0.7875 13.9C0.2625 12.6833 0 11.3833 0 10C0 8.61667 0.2625 7.31667 0.7875 6.1C1.3125 4.88333 2.025 3.825 2.925 2.925C3.825 2.025 4.88333 1.3125 6.1 0.7875C7.31667 0.2625 8.61667 0 10 0C11.3833 0 12.6833 0.2625 13.9 0.7875C15.1167 1.3125 16.175 2.025 17.075 2.925C17.975 3.825 18.6875 4.88333 19.2125 6.1C19.7375 7.31667 20 8.61667 20 10C20 11.3833 19.7375 12.6833 19.2125 13.9C18.6875 15.1167 17.975 16.175 17.075 17.075C16.175 17.975 15.1167 18.6875 13.9 19.2125C12.6833 19.7375 11.3833 20 10 20ZM10 18C12.2333 18 14.125 17.225 15.675 15.675C17.225 14.125 18 12.2333 18 10C18 7.76667 17.225 5.875 15.675 4.325C14.125 2.775 12.2333 2 10 2C7.76667 2 5.875 2.775 4.325 4.325C2.775 5.875 2 7.76667 2 10C2 12.2333 2.775 14.125 4.325 15.675C5.875 17.225 7.76667 18 10 18Z" fill="#FF4949"></path>
</svg>
</div>"
`;

exports[`when icon is set close-able > should render origin icon when double clicked 1`] = `
"<div id="icon-base-test-10" class="_icon-base_2178eb _icon-base-dark_2178eb"><svg id="test"></svg>
</div>"
`;

exports[`when set elementTargetID > should render if element id exist 1`] = `
"<div id="icon-base-test-8" class="_icon-base_2178eb"><svg width="16" height="21" viewBox="0 0 16 21" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M2 21C1.45 21 0.979167 20.8042 0.5875 20.4125C0.195833 20.0208 0 19.55 0 19V9C0 8.45 0.195833 7.97917 0.5875 7.5875C0.979167 7.19583 1.45 7 2 7H3V5C3 3.61667 3.4875 2.4375 4.4625 1.4625C5.4375 0.4875 6.61667 0 8 0C9.38333 0 10.5625 0.4875 11.5375 1.4625C12.5125 2.4375 13 3.61667 13 5V7H14C14.55 7 15.0208 7.19583 15.4125 7.5875C15.8042 7.97917 16 8.45 16 9V19C16 19.55 15.8042 20.0208 15.4125 20.4125C15.0208 20.8042 14.55 21 14 21H2ZM2 19H14V9H2V19ZM8 16C8.55 16 9.02083 15.8042 9.4125 15.4125C9.80417 15.0208 10 14.55 10 14C10 13.45 9.80417 12.9792 9.4125 12.5875C9.02083 12.1958 8.55 12 8 12C7.45 12 6.97917 12.1958 6.5875 12.5875C6.19583 12.9792 6 13.45 6 14C6 14.55 6.19583 15.0208 6.5875 15.4125C6.97917 15.8042 7.45 16 8 16ZM5 7H11V5C11 4.16667 10.7083 3.45833 10.125 2.875C9.54167 2.29167 8.83333 2 8 2C7.16667 2 6.45833 2.29167 5.875 2.875C5.29167 3.45833 5 4.16667 5 5V7Z" fill="#FEF7F7"></path>
</svg>
</div>"
`;

exports[`when set theme > should render the right theme 1`] = `
"<div id="icon-base-test-5" class="_icon-base_2178eb"><svg width="16" height="21" viewBox="0 0 16 21" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M2 21C1.45 21 0.979167 20.8042 0.5875 20.4125C0.195833 20.0208 0 19.55 0 19V9C0 8.45 0.195833 7.97917 0.5875 7.5875C0.979167 7.19583 1.45 7 2 7H3V5C3 3.61667 3.4875 2.4375 4.4625 1.4625C5.4375 0.4875 6.61667 0 8 0C9.38333 0 10.5625 0.4875 11.5375 1.4625C12.5125 2.4375 13 3.61667 13 5V7H14C14.55 7 15.0208 7.19583 15.4125 7.5875C15.8042 7.97917 16 8.45 16 9V19C16 19.55 15.8042 20.0208 15.4125 20.4125C15.0208 20.8042 14.55 21 14 21H2ZM2 19H14V9H2V19ZM8 16C8.55 16 9.02083 15.8042 9.4125 15.4125C9.80417 15.0208 10 14.55 10 14C10 13.45 9.80417 12.9792 9.4125 12.5875C9.02083 12.1958 8.55 12 8 12C7.45 12 6.97917 12.1958 6.5875 12.5875C6.19583 12.9792 6 13.45 6 14C6 14.55 6.19583 15.0208 6.5875 15.4125C6.97917 15.8042 7.45 16 8 16ZM5 7H11V5C11 4.16667 10.7083 3.45833 10.125 2.875C9.54167 2.29167 8.83333 2 8 2C7.16667 2 6.45833 2.29167 5.875 2.875C5.29167 3.45833 5 4.16667 5 5V7Z" fill="#FEF7F7"></path>
</svg>
</div>"
`;

exports[`when set theme > should render the right theme 2`] = `
"<div id="icon-base-test-5" class="_icon-base_2178eb"><svg width="16" height="21" viewBox="0 0 16 21" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M2 21C1.45 21 0.979167 20.8042 0.5875 20.4125C0.195833 20.0208 0 19.55 0 19V9C0 8.45 0.195833 7.97917 0.5875 7.5875C0.979167 7.19583 1.45 7 2 7H3V5C3 3.61667 3.4875 2.4375 4.4625 1.4625C5.4375 0.4875 6.61667 0 8 0C9.38333 0 10.5625 0.4875 11.5375 1.4625C12.5125 2.4375 13 3.61667 13 5V7H14C14.55 7 15.0208 7.19583 15.4125 7.5875C15.8042 7.97917 16 8.45 16 9V19C16 19.55 15.8042 20.0208 15.4125 20.4125C15.0208 20.8042 14.55 21 14 21H2ZM2 19H14V9H2V19ZM8 16C8.55 16 9.02083 15.8042 9.4125 15.4125C9.80417 15.0208 10 14.55 10 14C10 13.45 9.80417 12.9792 9.4125 12.5875C9.02083 12.1958 8.55 12 8 12C7.45 12 6.97917 12.1958 6.5875 12.5875C6.19583 12.9792 6 13.45 6 14C6 14.55 6.19583 15.0208 6.5875 15.4125C6.97917 15.8042 7.45 16 8 16ZM5 7H11V5C11 4.16667 10.7083 3.45833 10.125 2.875C9.54167 2.29167 8.83333 2 8 2C7.16667 2 6.45833 2.29167 5.875 2.875C5.29167 3.45833 5 4.16667 5 5V7Z" fill="#FEF7F7"></path>
</svg>
</div><div id="icon-base-test-6" class="_icon-base_2178eb _icon-base-dark_2178eb"><svg width="16" height="21" viewBox="0 0 16 21" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M2 21C1.45 21 0.979167 20.8042 0.5875 20.4125C0.195833 20.0208 0 19.55 0 19V9C0 8.45 0.195833 7.97917 0.5875 7.5875C0.979167 7.19583 1.45 7 2 7H3V5C3 3.61667 3.4875 2.4375 4.4625 1.4625C5.4375 0.4875 6.61667 0 8 0C9.38333 0 10.5625 0.4875 11.5375 1.4625C12.5125 2.4375 13 3.61667 13 5V7H14C14.55 7 15.0208 7.19583 15.4125 7.5875C15.8042 7.97917 16 8.45 16 9V19C16 19.55 15.8042 20.0208 15.4125 20.4125C15.0208 20.8042 14.55 21 14 21H2ZM2 19H14V9H2V19ZM8 16C8.55 16 9.02083 15.8042 9.4125 15.4125C9.80417 15.0208 10 14.55 10 14C10 13.45 9.80417 12.9792 9.4125 12.5875C9.02083 12.1958 8.55 12 8 12C7.45 12 6.97917 12.1958 6.5875 12.5875C6.19583 12.9792 6 13.45 6 14C6 14.55 6.19583 15.0208 6.5875 15.4125C6.97917 15.8042 7.45 16 8 16ZM5 7H11V5C11 4.16667 10.7083 3.45833 10.125 2.875C9.54167 2.29167 8.83333 2 8 2C7.16667 2 6.45833 2.29167 5.875 2.875C5.29167 3.45833 5 4.16667 5 5V7Z" fill="#FEF7F7"></path>
</svg>
</div>"
`;
